src/VsExtension/VsEvents/OnBuildPackageRestorer.cs

```diff
@@ -201,42 +208,87 @@ private void BuildEvents_OnBuildBegin(vsBuildScope scope, vsBuildAction Action)
+                            ThreadedWaitDialogProgress = threadedWaitDialogSession.Progress;
 
-                        // Restore packages and create the lock file for each project
-                        foreach (var project in buildEnabledProjects)
-                        {
-                            var projectName = NuGetProject.GetUniqueNameOrName(project);
-                            await BuildIntegratedProjectRestoreAsync(project, loggingProjectContext, enabledSources, Token);
-                            WriteLine(VerbosityLevel.Normal, Resources.PackageRestoreFinishedForProject, projectName);
-                        }
+                            // Restore packages and create the lock file for each project
+                            foreach (var project in buildEnabledProjects)
+                            {
+                                var projectName = NuGetProject.GetUniqueNameOrName(project);
+                                await BuildIntegratedProjectRestoreAsync(project, loggingProjectContext, enabledSources, Token);
+                                WriteLine(VerbosityLevel.Normal, Resources.PackageRestoreFinishedForProject, projectName);
+                            }
 
-                        WriteLine(canceled: Canceled, hasMissingPackages: true, hasErrors: HasErrors);
+                            WriteLine(canceled: Canceled, hasMissingPackages: true, hasErrors: HasErrors);
+                        }
                     }
                 }
             }
         }
 
+        /// <summary>
+        /// Determine if a full restore is required. For scenarios with no floating versions
+        /// where all packages are already on disk and no server requests are needed
+        /// we can skip the restore after some validation checks.
+        /// </summary>
+        private async Task<bool> IsRestoreRequired(List<BuildIntegratedProjectSystem> projects)
+        {
+            try
+            {
+                // Swap caches 
+                var oldCache = _buildIntegratedCache;
+                _buildIntegratedCache = await BuildIntegratedRestoreUtility.CreateBuildIntegratedProjectStateCache(projects);
+
+                if (BuildIntegratedRestoreUtility.CacheHasChanges(oldCache, _buildIntegratedCache))
+                {
+                    // A new project has been added
+                    return true;
+                }
+
+                var globalPackagesFolder = SettingsUtility.GetGlobalPackagesFolder(Settings);
+                var pathResolver = new VersionFolderPathResolver(globalPackagesFolder);
+
+                if (BuildIntegratedRestoreUtility.IsRestoreRequired(projects, pathResolver))
+                {
+                    // The project.json file does not match the lock file
+                    return true;
+                }
+            }
+            catch
+            {
+                Debug.Fail("Unable to validate lock files.");
```

Comment:
I've added writing to the output under diagnostic mode, and logging to the activity log for this.